blockchain/difficulty.go

```diff
@@ -743,19 +748,277 @@ func (b *BlockChain) calcNextRequiredStakeDifficulty(curNode *blockNode) (int64,
 	return nextDiff, nil
 }
 
-// CalcNextRequiredStakeDifficulty is the exported version of the above function.
-// This function is NOT safe for concurrent access.
+// estimateSupply returns an estimate of the coin supply for the provided block
+// height.  This is primarily used in the stake difficulty algorithm and relies
+// on an estimate to simplify the necessary calculations.  The actual total
+// coin supply as of a given block height depends on many factors such as the
+// number of votes included in every prior block (not including all votes
+// reduces the subsidy) and whether or not any of the prior blocks have been
+// invalidated by stakeholders thereby removing the PoW subsidy for them.
+//
+// This function is safe for concurrent access.
+func estimateSupply(params *chaincfg.Params, height int64) int64 {
+	if height <= 0 {
+		return 0
+	}
+
+	// Estimate the supply by calculating the full block subsidy for each
+	// reduction interval and multiplying it the number of blocks in the
+	// interval then adding the subsidy produced by number of blocks in the
+	// current interval.
+	supply := params.BlockOneSubsidy()
+	reductions := int64(height) / params.SubsidyReductionInterval
+	subsidy := params.BaseSubsidy
+	for i := int64(0); i < reductions; i++ {
+		supply += params.SubsidyReductionInterval * subsidy
+
+		subsidy *= params.MulSubsidy
+		subsidy /= params.DivSubsidy
+	}
+	supply += (1 + int64(height)%params.SubsidyReductionInterval) * subsidy
+
+	// Blocks 0 and 1 have special subsidy amounts that have already been
+	// added above, so remove what their subsidies would have normally been
+	// which were also added above.
+	supply -= params.BaseSubsidy * 2
+
+	return supply
+}
+
+// sumPurchasedTickets returns the sum of the number of tickets purchased in the
+// most recent specified number of blocks from the point of view of the passed
+// node.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) sumPurchasedTickets(startNode *blockNode, numToSum int64) (int64, error) {
+	var numPurchased int64
+	for node, numTraversed := startNode, int64(0); node != nil &&
+		numTraversed < numToSum; numTraversed++ {
+
+		numPurchased += int64(node.header.FreshStake)
+
+		// Get the previous block node.  This function is used over
+		// simply accessing iterNode.parent directly as it will
+		// dynamically create previous block nodes as needed.  This
+		// helps allow only the pieces of the chain that are needed
+		// to remain in memory.
+		var err error
+		node, err = b.getPrevNodeFromNode(node)
+		if err != nil {
+			return 0, err
```

Comment:
It's correct.  The only way this can fail is due to failing to load the node from the database and if that happens something is very wrong and there is no way to continue, so it must error.  That is especially critical since this is consensus code, it can't just ignore failure to retrieve the data.